src/vcpkg/binarycaching.cpp

```diff
@@ -952,68 +952,71 @@ namespace
 
-        std::string lookup_cache_entry(const std::string& abi) const
-        {
-            auto cmd = command()
-                           .string_arg(m_read_url)
-                           .string_arg("-G")
-                           .string_arg("-d")
-                           .string_arg("keys=vcpkg")
-                           .string_arg("-d")
-                           .string_arg("version=" + abi);
+            auto res = invoke_http_request(
+                "GET", std::vector<std::string>{m_content_type_header, m_token_header, m_accept_header}, url);
+
+            auto maybe_json = Json::parse_object(res.get()->c_str());
+            if (auto json = maybe_json.get())
+            {
+                auto archive_location = json->get("archiveLocation");
+                if (archive_location && archive_location->is_string())
+                {
+                    return archive_location->string(VCPKG_LINE_INFO).to_string();
+                }
+            }
 
-            std::vector<std::string> lines;
-            auto res = cmd_execute_and_capture_output(cmd);
-            if (!res.has_value() || res.get()->exit_code) return {};
-            auto json = Json::parse_object(res.get()->output);
-            if (!json.has_value() || !json.get()->contains("archiveLocation")) return {};
-            return json.get()->get("archiveLocation")->string(VCPKG_LINE_INFO).to_string();
+            return {};
         }
 
-        Optional<int64_t> reserve_cache_entry(const std::string& abi, int64_t cacheSize) const
+        Optional<int64_t> reserve_cache_entry(const std::string& package_name,
+                                              const std::string& abi,
+                                              int64_t cacheSize) const
         {
             Json::Object payload;
-            payload.insert("key", "vcpkg");
+            payload.insert("key", package_name + "-" + abi);
             payload.insert("version", abi);
             payload.insert("cacheSize", Json::Value::integer(cacheSize));
-            auto cmd = command().string_arg(m_write_url).string_arg("-d").string_arg(stringify(payload));
 
-            auto res = cmd_execute_and_capture_output(cmd);
-            if (!res.has_value() || res.get()->exit_code) return {};
-            auto json = Json::parse_object(res.get()->output);
-            if (!json.has_value() || !json.get()->contains("cacheId")) return {};
-            return json.get()->get("cacheId")->integer(VCPKG_LINE_INFO);
+            auto res =
+                invoke_http_request("POST",
+                                    std::vector<std::string>{m_content_type_header, m_token_header, m_accept_header},
+                                    m_write_url,
+                                    std::string{stringify(payload)});
+
+            auto maybe_json = Json::parse_object(res.get()->c_str());
```

Comment:
You haven't tested `res` for failure -- this is a potential null pointer dereference (UB).

You need to have another layer of:
```cpp
if (auto p = res.get()) {
   // use p->c_str()
```

But also -- `parse_object()` should not require `->c_str()`. You should be able to just `Json::parse_object(*p)`.

This is all just another thing you have to constantly consider when writing C++ -- _anywhere_ you do `*x` or `x->`, you MUST ask yourself "How do I know for sure that `x` is not null here?". Typically, we get proof of that by testing `x`, either as `if (x) {....}` or `if (auto x = <expr>) {....}`.